### ISSUE TITLE
distances.c calloc for zero objects gives SIGILL

### ISSUE THREAD

**Issue by ompiteam** [open-mpi/ompi#190]
A call to calloc is made for zero objects. Result is SIGILL.

stack from gdb -- see position 6:

```
#0  0x00007ffff5a26b67 in kill () from /lib64/libc.so.6
#1  0x00007ffff65c4435 in ?? () from /usr/lib64/libefence.so.0
#2  0x00007ffff65c47aa in EF_Abortv () from /usr/lib64/libefence.so.0
#3  0x00007ffff65c483c in EF_Abort () from /usr/lib64/libefence.so.0
#4  0x00007ffff65c4009 in memalign () from /usr/lib64/libefence.so.0
#5  0x00007ffff65c4275 in calloc () from /usr/lib64/libefence.so.0
#6  0x00007ffff68b84c6 in opal_hwloc132_hwloc_convert_distances_indexes_into_objects (
    topology=topology@entry=0x7ffff224a000) at distances.c:289
#7  0x00007ffff69067ca in hwloc_discover (topology=0x7ffff224a000) at topology.c:2023
#8  opal_hwloc132_hwloc_topology_load (topology=0x7ffff224a000) at topology.c:2596
#9  0x00007ffff68c0fd7 in opal_hwloc_unpack (buffer=0x7ffff1bb2000, dest=<optimized out>, num_vals=0x7fffffffd010,
    type=<optimized out>) at base/hwloc_base_dt.c:83
#10 0x00007ffff68bc7ee in opal_dss_unpack_buffer (buffer=buffer@entry=0x7ffff1bb2000,
    dst=dst@entry=0x7ffff6b62b08 <opal_hwloc_topology>, num_vals=num_vals@entry=0x7fffffffd010,
    type=type@entry=22 '\026') at dss/dss_unpack.c:120
#11 0x00007ffff68bd79a in opal_dss_unpack (buffer=0x7ffff1bb2000, dst=0x7ffff6b62b08 <opal_hwloc_topology>,
    num_vals=0x7fffffffd080, type=22 '\026') at dss/dss_unpack.c:84
#12 0x00007ffff68853ff in orte_util_nidmap_init (buffer=0x7ffff1bb2000) at util/nidmap.c:146
#13 0x00007ffff150b6fa in rte_init () at ess_env_module.c:173
#14 0x00007ffff686e4da in orte_init (pargc=pargc@entry=0x0, pargv=pargv@entry=0x0, flags=flags@entry=32)
    at runtime/orte_init.c:127
#15 0x00007ffff6830899 in ompi_mpi_init (argc=argc@entry=0, argv=argv@entry=0x0, requested=0, provided=0x7fffffffd310)
    at runtime/ompi_mpi_init.c:357
#16 0x00007ffff6846ad6 in PMPI_Init (argc=0x0, argv=0x0) at pinit.c:86
#17 0x00000000004156c1 in MPI::Init () at /usr/local/include/openmpi/ompi/mpi/cxx/functions_inln.h:128
#18 0x00000000004135c2 in main () at Test_SimOutputSpatialNc.cpp:211
```

Examine stack position 6 shows call calloc(0, ...)

```
(gdb) f 6
#6  0x00007ffff68b84c6 in opal_hwloc132_hwloc_convert_distances_indexes_into_objects (
    topology=topology@entry=0x7ffff224a000) at distances.c:289
289           hwloc_obj_t *objs = calloc(nbobjs, sizeof(hwloc_obj_t));
(gdb) p nbobjs
$3 = 0
(gdb) l
284         unsigned nbobjs = topology->os_distances[type].nbobjs;
285         unsigned *indexes = topology->os_distances[type].indexes;
286         float *distances = topology->os_distances[type].distances;
287         unsigned i, j;
288         if (!topology->os_distances[type].objs) {
289           hwloc_obj_t *objs = calloc(nbobjs, sizeof(hwloc_obj_t));
```

Possible fix: insert after line 284: if (nbobjs == 0) return;

```
Platform: opensuse 13.1 64-bit
g++ version: 4.8.1 20130909 [gcc-4_8-branch revision 202388]
configure options:  CFLAGS='-ggdb3 -fPIC'
```

**Comment by bgoglin** [comment]
FYI, this was fixed in hwloc 1.11.3 in early 2016 but all branches < 3.0.x still have 1.11.2 (with a couple unrelated patchs).

**Comment by jsquyres** [comment]
This doesn't appear to have been a major problem since 2014, and it will be fixed in >v3.0.0.  So I think we should close this issue.
